conda_libmamba_solver/solver.py

```diff
@@ -364,12 +365,13 @@ def _solve_attempt(
         log.debug("Computed specs: %s", out_state.specs)
 
         # ## Convert to tasks
-        out_state.pins.clear()
         n_pins = 0
         tasks = self._specs_to_tasks(in_state, out_state)
         for (task_name, task_type), specs in tasks.items():
             log.debug("Adding task %s with specs %s", task_name, specs)
-            if task_name == "ADD_PIN":
+            if task_name == "ADD_PIN" and attempt == 1:
```

Comment:
Makes sense given the Matrix discussion linked in the OP.

If it doesn't complicate things too much, would it make more sense to add an `is_initial_attempt: bool` (or just `attempt: int`) parameter to `_specs_to_tasks` and have the `ADD_PIN` task simply not created?
To me, that would conceptually make more sense since creating a "task" that will be ignored otherwise is awkward ;).

Reply:
(Or the tasks could be created beforehand and split into initial and recurring ones; but that might need more restructuring which might be overkill for this single special-handling case for now.)

Reply:
Yea I get your point, but here's my rationale:

IMO the limitation that `solver.add_pins()` has persistent effects in the pool is an implementation detail of the libmamba library. `specs_to_tasks` aims to be as unaware of those implementation details as possible. It's not that we shouldn't pin in the 2nd iteration and following ones; we do want, but there's a technical limitation. I'd like `specs_to_tasks` to reflect that logic, hence why I ignore it here.

It could be cleaner, but we are going to remove a lot of code from the `state, and I plan to include some cleanup in the solver module.

I hope it makes sense.

Reply:
Yes, thanks!

Reply:
How many attempts would be typical?

Reply:
Depends on the environment. Hopefully the first one is enough. But with old ones with many war scars, it can take a few. We fast-track some cases, like changing Python version, because otherwise we'd need for the solver to realize all the Python-depending packages are conflicting.